realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/RealmFunctionsTests.kt

```diff
@@ -116,36 +162,205 @@ class RealmFunctionsTests {
+                assertTrue(result.error.exception is CodecConfigurationException)
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    @Test
+    fun localCodecRegistry() {
+        val input = Dog("PojoFido")
+        assertEquals(input, functions.callFunction(FIRST_ARG_FUNCTION, listOf(input), Dog::class.java, pojoRegistry))
+    }
 
+    @Test
+    fun asyncLocalCodecRegistry() = looperThread.runBlocking {
+        val input = Dog("PojoFido")
+        functions.callFunctionAsync(FIRST_ARG_FUNCTION, listOf(input), Dog::class.java, pojoRegistry) { result ->
+            if (result.isSuccess) {
+                assertEquals(input, result.get())
+            } else  {
+                fail()
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    @Test
+    fun instanceCodecRegistry() {
+        val input = Dog("PojoFido")
+        val functionsWithCodecRegistry = anonUser.getFunctions(pojoRegistry)
+        assertEquals(input, functionsWithCodecRegistry.callFunction(FIRST_ARG_FUNCTION, listOf(input), Dog::class.java))
+    }
+
+    @Test
+    fun unknownFunction() {
+        assertFailsWith<ObjectServerError> {
+            functions.callFunction("unknown", listOf(32), Dog::class.java)
+        }
+    }
+
+    @Test
+    fun asyncUnknownFunction() = looperThread.runBlocking {
         val input = Dog("PojoFido")
+        functions.callFunctionAsync("unknown", listOf(input), Dog::class.java, pojoRegistry) { result ->
+            if (result.isSuccess) {
+                fail()
+            } else  {
+                // FIXME How verify exact error. NativeErrorIntValue? Or error message?
+                assertTrue(result.error is ObjectServerError)
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    @Test
+    fun asyncNonLoopers() {
+        assertFailsWith<IllegalStateException> {
+            functions.callFunctionAsync(FIRST_ARG_FUNCTION, listOf(32), Integer::class.java, pojoRegistry) { result ->
+                fail()
+            }
+        }
+    }
+
+    @Test
+    fun callFunction_sum() {
+        val numbers = listOf(1, 2, 3, 4)
+        assertEquals(10, functions.callFunction("sum", numbers, Integer::class.java).toInt())
+    }
 
-        val actual: Dog = functions.invoke(input, Dog::class.java, pojoRegistry)
+    @Test
+    fun callFunction_remoteError() {
+        assertFailsWith<ObjectServerError> {
+            // FIXME Do we need to assert more about the error
+            functions.callFunction("error", emptyList<Any>(), String::class.java)
+        }
+    }
+
+    @Test
+    fun callFunction_null() {
+        assertTrue(functions.callFunction("null", emptyList<Any>(), BsonNull::class.java).isNull)
+    }
+
+    @Test
+    fun callFunction_empty() {
+        assertEquals(BsonType.UNDEFINED, functions.callFunction("empty", emptyList<Any>(), BsonUndefined::class.java).bsonType)
+    }
+
+    // FIXME Do we need to test connectivity issues
```

Comment:
No, we normally don't do that. The general network transport is tested elsewhere and as long as there are tests that ensure an exception/error handler can be correctly called we assume that connection errors will also be reported correctly.